### PR TITLE
feat: make grpc insert requests in a batch

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Trying to optimize Prometheus remote write.

Now the remote write path is like this:

```txt
remote write     ->    grpc insert     ->   `DistInserter`(splits inserts here)      -> Datanodes
(single request)     (single request)       (multiple batching requests)                (parallel)
                                            (other protocols insertions all go here)
```

Before:

```txt
remote write     ->    grpc insert     ->   `DistTable::insert`(splits inserts here) -> Datanodes
(single request)     (multiple requests,    (single request)                         (single request)
                      for each table)
```

This PR also:

- makes the inserts in Datanode GRPC interface parallel.
- makes the table "create-on-insert" parallel (we have to add a distributed lock in Metasrv)
  - will be changed to use procedure soon
- makes the gRPC server no longer blocks

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by MichaelScofield** [comment]
@fengjiachun PTAL

**Review by fengjiachun — APPROVED** [review]

File: src/common/grpc/src/channel_manager.rs [hunk 1]
```diff
@@ -28,6 +28,7 @@ use tower::make::MakeConnection;
 use crate::error::{CreateChannelSnafu, InvalidConfigFilePathSnafu, InvalidTlsConfigSnafu, Result};
 
 const RECYCLE_CHANNEL_INTERVAL_SECS: u64 = 60;
+const DEFAULT_REQUEST_TIMEOUT_SECS: u64 = 2;
```
Comment (killme2008) [inline src/common/grpc/src/channel_manager.rs:31]: Maybe too short, I think 5~10 seconds is more reasonable. Is it configurable?

File: src/common/meta/src/rpc/router.rs [hunk 2]
```diff
@@ -266,6 +267,13 @@ impl TableRoute {
             })
             .collect()
     }
+
+    pub fn find_region_leader(&self, region_number: RegionNumber) -> Option<Peer> {
```
Comment (killme2008) [inline src/common/meta/src/rpc/router.rs:271]: This function may be inefficient when route info is large, I think we can build a map that keeps `<region_number, Peer>` when route info is building.

File: src/datanode/src/instance/grpc.rs [hunk 3]
```diff
@@ -119,31 +121,47 @@ impl Instance {
...
+                        &table_name,
+                    ),
+                })
+            });
+        }
 
-        let affected_rows = table.insert(request).await.with_context(|_| InsertSnafu {
-            table_name: table_ref.to_string(),
-        })?;
+        let results = future::try_join_all(inserts.into_iter().map(common_runtime::spawn_write))
```
Comment (killme2008) [inline src/datanode/src/instance/grpc.rs:161]: Why not `spawn_write` above? Then the `inserts` is the vector of futures to join.

File: src/frontend/src/instance/distributed.rs [hunk 4]
```diff
@@ -569,36 +584,21 @@ impl DistInstance {
...
+        let inserter = DistInserter::new(
+            ctx.current_catalog(),
+            ctx.current_schema(),
+            self.catalog_manager.clone(),
+        );
+        let affected_rows = inserter.grpc_insert(requests).await?;
+        Ok(Output::AffectedRows(affected_rows as usize))
     }
 
+    // TODO(LFC): Like insertions above, refactor GRPC deletion impl here.
```
Comment (killme2008) [inline src/frontend/src/instance/distributed.rs:601]: There are a lot of `TODO` in this PR, we can create some issues to track them.

File: src/frontend/src/table/insert.rs [hunk 5]
```diff
@@ -120,23 +70,136 @@ pub(crate) fn to_grpc_insert_request(
     })
 }
 
+fn vector_to_grpc_column(
+    table_meta: &TableMeta,
+    column_name: &str,
+    vector: VectorRef,
+) -> Result<Column> {
+    // Safety: timestamp column has to be existed in a table.
+    let semantic_type = if column_name == table_meta.schema.timestamp_column().unwrap().name {
```
Comment (killme2008) [inline src/frontend/src/table/insert.rs:79]: We have external tables and maybe other table engines in future. So the `unwrap` is a trap in the future. I think we can add an `ensure!` to return an error.

**Review by Fengys123 — APPROVED** [review]

File: src/servers/src/grpc.rs [hunk 6]
```diff
@@ -94,6 +101,22 @@ impl GrpcServer {
...
     }
+
+    pub async fn wait_for_serve(&self) -> Result<()> {
+        let mut serve_state = self.serve_state.lock().await;
+        let rx = serve_state.take().context(InternalSnafu {
+            err_msg: "gRPC serving state is unknown, maybe the server is not started, \
+                      or we have already waited for the serve result before.",
+        })?;
+        let Ok(result) = rx.await else {
+            warn!("Background gRPC serving task is quit before we can receive the serve result.");
```
Comment (killme2008) [inline src/servers/src/grpc.rs:112]:
```suggestion
            warn!("Background gRPC serving task quited before we can receive the serve result.");
```

File: src/common/grpc/src/channel_manager.rs (hunk 1 above)
Comment (MichaelScofield) [reply src/common/grpc/src/channel_manager.rs:31]: yes, set to 8 seconds when creating clients for datanodes in frontend

File: src/frontend/src/instance/distributed.rs (hunk 4 above)
Comment (MichaelScofield) [reply src/frontend/src/instance/distributed.rs:601]: will find some time to do that for all my todos

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**